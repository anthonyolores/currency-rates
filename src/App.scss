
$bd-radius:0px !important;
$primary-color: #240040;
%form-style{
  padding: .375rem .75rem;
  border-radius: $bd-radius;
  border-width: 1px;
  vertical-align: middle;
  width:100%;
  color:#6b00bc;
  &:focus {
    outline: none !important;
    border:1px solid #6D00BF;
    box-shadow: none;
  }
  &:disabled{
    background:#eee8f3;
    border: 1px solid #eee8f3 !important;
  }
}

input, button, select, .list-group-item{
  @extend %form-style;
}
input{
  border:1px solid #e5e5e5;
}
.header-container{
  padding: 20px 30px 10px !important;
  background: $primary-color;
  background: linear-gradient(90deg, rgba(36,0,64,1) 0%, rgb(66, 0, 117) 100%);
  h1{
    color:#fff;
    font-size:1.5em;
    margin-bottom: 0px;
  }
  p{
    font-size:0.8em;
    color:#f9f2ff;
    opacity: 0.6;
    margin-bottom: 5px;
  }
}
.form-container{
  background: #fff;
  padding: 20px 30px !important;
  box-shadow: 0 1px 3px rgba(25,25,25,.1);
}
.list-container{
  position: absolute;
  z-index: 1;
  width: 100%;
  padding-left: 0px;
  // padding-right: 10px;
  max-height: 261px;
  overflow: auto;
  border-bottom: 1px solid #eee;
  cursor: pointer;
}

body{
  margin-top:25px;
  background:#eee;
}

@media only screen and (max-width:575px){
  .container{
    .row{
      margin:0px !important;
    }
  }
}