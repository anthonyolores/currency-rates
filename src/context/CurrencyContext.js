
import React, { createContext } from 'react';

export const context = {
};
export const CurrencyContext = createContext(context);