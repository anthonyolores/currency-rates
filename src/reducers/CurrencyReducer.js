
export const initialState = {
    currency: null,
    fromNumber: 0,
    fromCurrency: '',
    toNumber:0,
    toCurrency:'',
    isFromCurrency: false,
    isToCurrency: false
  };

export const actions = {
    SET_CURRENCY: 'SET_CURRENCY',
    SET_FROM_NUMBER: 'SET_FROM_NUMBER',
    SET_FROM_CURRENCY: 'SET_FROM_CURRENCY',
    SET_TO_NUMBER: 'SET_TO_NUMBER',
    SET_TO_CURRENCY: 'SET_TO_CURRENCY',
    SET_IS_FROM_CURRENCY: 'SET_IS_FROM_CURRENCY',
    SET_IS_TO_CURRENCY: 'SET_IS_TO_CURRENCY'
  };
  
export const reducer = (state, action) => {
  console.log(action.type + '---' + action.data);
    switch(action.type){
      case actions.SET_CURRENCY: {
        if(action.data && action.data.rates != null){
          const listRates = Object.keys(action.data.rates);
          action.data.rates = [
            {value: action.data.rates[listRates[0]], currency: listRates[0]},
            {value: action.data.rates[listRates[1]], currency: listRates[1]}
          ]
        }
        return {
          ...state,
          currency: action.data
        }
      }
      case actions.SET_FROM_NUMBER: {
        return {
          ...state,
          fromNumber: action.data
        }     
      }
      case actions.SET_FROM_CURRENCY: {
        return {
          ...state,
          fromCurrency: action.data
        }      
      }
      case actions.SET_TO_NUMBER: {
        return {
          ...state,
          toNumber: action.data
        }      
      }
      case actions.SET_TO_CURRENCY: {
        return {
          ...state,
          toCurrency: action.data
        }      
      }
      case actions.SET_IS_FROM_CURRENCY: {
        return {
          ...state,
          isFromCurrency: action.data==null?false:action.data,
          isToCurrency: action.data==null?false:!action.data
        }      
      }
      case actions.SET_IS_TO_CURRENCY: {
        return {
          ...state,
          isFromCurrency: action.data==null?false:!action.data,
          isToCurrency: action.data==null?false:action.data
        }      
      }
      default:{
        return initialState;
      }
    }
  };