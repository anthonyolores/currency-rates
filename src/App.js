import './App.scss';
import React, { useReducer, useContext, useRef, useEffect, useState, useCallback, useMemo} from 'react';
import CmpInputText from './components/InputText';
import {Container, Row, Col, ListGroup} from 'react-bootstrap';
import {CurrencyContext} from '../src/context/CurrencyContext';
import {actions, reducer, initialState} from './reducers/CurrencyReducer';
import CurrencyService from './services/CurrencyService';
import {Chart} from 'chart.js';
const colStyle = {
  padding:5
};
const abbrs = CurrencyService.getAbbr();

function filterAbbrList(keyword){
  let tempList = [];
  if(keyword === ''){
    tempList = abbrs.slice();
  }
  else{
    keyword = keyword.toUpperCase();
    for(let i = 0; i < abbrs.length; i++){
      if(abbrs[i].indexOf(keyword) !== -1){
        tempList.push(abbrs[i]);
      }
    }
  }
  return tempList;
}

function isCurrencyValid(currency){
  if(currency && currency.length > 0){
    return abbrs.indexOf(currency.toUpperCase()) !== -1?true:false;  
  }
  else{
    return false;
  } 
}

function getResult(num, currency, operation){
  let result  = 0;
  if(num != ''){
    num = parseFloat(num);
    const isLess0 = (currency.rates[1].value<0);
    const rate = currency.rates[1].value == 1? currency.rates[0].value : currency.rates[1].value;
    if(operation === 0){
      result = isLess0? (num / rate) : (num * rate);
    }
    else{
      result = isLess0? (num * rate) : (num / rate);
    }
  }

  return result % 1 != 0? result.toFixed(2) : result;
}

function setCurrency(from, to, dispatch){
  if(isCurrencyValid(from) && isCurrencyValid(to)){
    if(from == to){
      dispatch({type:actions.SET_CURRENCY, data:{
        base: 'from',
        date: '',
        rates:{
          [from]:1,
          [to+'2']:1,
        }
      }});
    }
    else{
      CurrencyService.getRates(from, to).then((response)=>{
        if(response.data.error){
          dispatch({type:actions.SET_CURRENCY, data:null});
        }
        else{
          dispatch({type:actions.SET_CURRENCY, data:response.data});    
        }
      }).catch((error)=>{
        dispatch({type:actions.SET_CURRENCY, data:null});
      });
    }
  }
  else{
    dispatch({type:actions.SET_CURRENCY, data:null});
  }
}

export const CurrencyList = ({actionType, actionType2}) => {
  const { currencyState, dispatch } = useContext(CurrencyContext);
  const listRef = useRef(null);
  //function must start with use if using hooks
  function useClickOutside(ref) {
    function handleClickOutside(event) {
      if (ref.current && 
        !ref.current.contains(event.target) && 
        event.target.placeholder !== 'Currency') {
        dispatch({type:actionType2, data:null});
      }
      else if(ref.current && ref.current.contains(event.target)){
        dispatch({type:actionType, data:event.target.innerText});
        dispatch({type:actionType2, data:null});
        setCurrency(currencyState.isFromCurrency?event.target.innerText:currencyState.fromCurrency, currencyState.isToCurrency?event.target.innerText:currencyState.toCurrency, dispatch);
      }
    }
    useEffect(() => {
      // Bind the event listener
      document.addEventListener("mousedown", handleClickOutside);

      return () => {
        // Unbind the event listener on clean up
        document.removeEventListener("mousedown", handleClickOutside);
      };
    });
  }

  useClickOutside(listRef);
  const abbrrList = filterAbbrList(currencyState.isFromCurrency?currencyState.fromCurrency:currencyState.toCurrency);  
  return <div className="list-container" ref={listRef}>
  <ListGroup defaultActiveKey="#link1">
      {
        abbrrList.map((item, index)=>{
          return <ListGroup.Item key={index}>{item}</ListGroup.Item>
        })
      }
  </ListGroup>
  </div>
  }

export const FromCurrency = () => {
  const { currencyState, dispatch } = useContext(CurrencyContext);
  const handler = useCallback((value) => {
    dispatch({type:actions.SET_FROM_CURRENCY, data:value});
    setCurrency(value, currencyState.toCurrency, dispatch);
  });
  const focusHandler = (value) => {
    dispatch({type:actions.SET_IS_FROM_CURRENCY, data:value});
  }
  return <CmpInputText value={currencyState.fromCurrency} placeholder={'Currency'} handler={handler} focusHandler={focusHandler}/>
}

export function ToCurrency(){
  const { currencyState, dispatch } = useContext(CurrencyContext);
  const handler = useCallback((value) => {
    dispatch({type:actions.SET_TO_CURRENCY, data:value});
    setCurrency(currencyState.fromCurrency, value, dispatch);
  });
  const focusHandler = (value) => {
    dispatch({type:actions.SET_IS_TO_CURRENCY, data:value});
  }
  return <CmpInputText value={currencyState.toCurrency} placeholder={'Currency'} handler={handler} focusHandler={focusHandler}/>
}

export function FromNumber(){
  const { currencyState, dispatch } = useContext(CurrencyContext);
  const [disable, setDisable] = useState(true);
  const handler = useCallback((value) => {
    if(/^[0-9.]*$/.test(value)){
      dispatch({type:actions.SET_FROM_NUMBER, data:value});  
      if((isCurrencyValid(currencyState.fromCurrency) && isCurrencyValid(currencyState.toCurrency))){
        dispatch({type:actions.SET_TO_NUMBER, data:getResult(value, currencyState.currency, 0)});
      }
    }
  });
  useEffect(()=>{
    setDisable(!(isCurrencyValid(currencyState.fromCurrency) && isCurrencyValid(currencyState.toCurrency)));
  }, [currencyState.fromCurrency, currencyState.toCurrency, currencyState.fromNumber, currencyState.toNumber]);
  return <CmpInputText value={currencyState.fromNumber} placeholder={0} handler={handler} disable={disable}/>
}

export function ToNumber(){
  const { currencyState, dispatch } = useContext(CurrencyContext);
  const [disable, setDisable] = useState(true);
  const handler = useCallback((value) => {
    if(/^[0-9.]*$/.test(value)){
      dispatch({type:actions.SET_TO_NUMBER, data:value});
      if((isCurrencyValid(currencyState.fromCurrency) && isCurrencyValid(currencyState.toCurrency))){
        dispatch({type:actions.SET_FROM_NUMBER, data:getResult(value, currencyState.currency, 1)});
      }
    }
  });
  useEffect(()=>{
    setDisable(!(isCurrencyValid(currencyState.fromCurrency) && isCurrencyValid(currencyState.toCurrency))); 
  }, [currencyState.fromCurrency, currencyState.toCurrency, currencyState.fromNumber, currencyState.toNumber]);
  return <CmpInputText value={currencyState.toNumber} placeholder={0} handler={handler} disable={disable}/>
}

function LineChart({color}){

  const { currencyState, dispatch } = useContext(CurrencyContext);
  const [lineChart, setLineChart] = useState(null);
  const [lineData, setLineData] = useState(null);
  const chartRef = useRef(null);

  useEffect(()=>{
    if((isCurrencyValid(currencyState.fromCurrency) && isCurrencyValid(currencyState.toCurrency))){
      CurrencyService.getHistory(currencyState.fromCurrency, currencyState.toCurrency).then((response)=>{
      //CurrencyService.getHistory("NZD", "PHP").then((response)=>{
        const dates = Object.keys(response.data.rates);
        const data = dates.map((item => {
          const currencies = Object.keys(response.data.rates[item]);
          const value = response.data.rates[item][currencies[0]] == 1? response.data.rates[item][currencies[1]] : response.data.rates[item][currencies[0]];
          return {label: item.replace('2019-',''), value: value};
        }));
        data.sort((a, b) => parseInt(a.label.replace('-', '')) - parseInt(b.label.replace('-', '')));
        setLineData(data);
        setLineChart(new Chart(chartRef.current, {
          type: 'line',
          data: {
            labels: data.map(d => d.label),
            datasets: [{
              label: 'Exchange Rate History',
              data: data.map(d => d.value),
              fill: 'none',
              backgroundColor: color,
              pointRadius: 4,
              borderColor: color,
              borderWidth: 1,
              lineTension: 0,
              steppedLine:false,
              
            }]
          }
        }));
        console.log(data);
      });
    }

  }, [currencyState.currency]);

  return lineData==null || currencyState.currency == null?''
    :
    <canvas
      ref={chartRef}
      data={lineData}
      title="My amazing data"
      color="#70CAD1"
    />
  
}

function App() {

  const [currencyState, dispatch] = useReducer(reducer, initialState);
  useEffect(()=>{
    if(currencyState.currency != null)
      dispatch({type:actions.SET_FROM_NUMBER, data:getResult(currencyState.toNumber, currencyState.currency, 1)});
  }, [currencyState.currency]);
  return useMemo(()=>{
    console.log(currencyState.fromCurrency + '-----' + currencyState.toCurrency);
    return (
      <CurrencyContext.Provider value={{ currencyState, dispatch }}>
        <div className="App">
          <Container>
            <Row className="justify-content-md-center justify-content-sm-center"> 
              <Col xs={12} sm={10} md={6} style={colStyle} className="header-container">   
                <Row>
                    <Col xs={12} sm={12} md={12} style={colStyle}>
                      <h1>CURRENCY</h1>
                      <p> Rates published by the European Central Bank</p>
                    </Col>
                  </Row>
              </Col>     
            </Row>
            <Row className="justify-content-md-center justify-content-sm-center">
              <Col xs={12} sm={10} md={6} style={colStyle} className="form-container">     
                <Row>
                  <Col xs={12} sm={6} md={6} style={colStyle}>
                    <FromCurrency />
                    {
                      currencyState.isFromCurrency?
                        <CurrencyList actionType={actions.SET_FROM_CURRENCY} actionType2={actions.SET_IS_FROM_CURRENCY}/>
                      :''
                    }           
                  </Col>
                  <Col xs={12} sm={6} md={6} style={colStyle}>
                    <FromNumber/>
                  </Col>
                </Row>
                <Row>
                  <Col xs={12} sm={6} md={6} style={colStyle}>
                    <ToCurrency />
                    {
                      currencyState.isToCurrency?
                        <CurrencyList actionType={actions.SET_TO_CURRENCY} actionType2={actions.SET_IS_TO_CURRENCY}/>
                      :''
                    }           
                  </Col>
                  <Col xs={12} sm={6} md={6} style={colStyle}>
                    <ToNumber/>
                  </Col>
                </Row>
                <Row>
                  <LineChart color="#6D00BF"/>
                </Row>
              </Col>
            </Row>
        </Container>
        </div>
      </CurrencyContext.Provider>
    );
  }, [currencyState, dispatch]);

}

export default App;
