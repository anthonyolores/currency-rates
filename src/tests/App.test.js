import CmpButton from '../components/Button';
import renderer from 'react-test-renderer';
import {render, fireEvent} from '@testing-library/react'

import {CurrencyContext} from '../context/CurrencyContext';
import React, { useReducer} from 'react';
import {actions, reducer, initialState} from '../reducers/CurrencyReducer';
import {FromCurrency} from '../App';

test('From currency snapshot', ()=>{
    const currencyState = initialState;
    const component = renderer.create(
        <CurrencyContext.Provider value={{currencyState}}>
             <FromCurrency />
        </CurrencyContext.Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
