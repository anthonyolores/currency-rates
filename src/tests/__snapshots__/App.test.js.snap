// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`From currency snapshot 1`] = `
<input
  onChange={[Function]}
  placeholder="Currency"
  type="text"
  value=""
/>
`;
